In [1]:
"I (We) certify that the code and data in this assignment were generated independently, using only the tools and resources defined in the course and that I (we) did not receive any external help, coaching or contributions during the production of this work."



'I (We) certify that the code and data in this assignment were generated independently, using only the tools and resources defined in the course and that I (we) did not receive any external help, coaching or contributions during the production of this work.'

In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

## Data Preprocessing

In [3]:
df = pd.read_csv("penguins.csv")

In [4]:
df.isnull().sum()

species               0
island                0
bill_length_mm        2
bill_depth_mm         2
flipper_length_mm     2
body_mass_g           2
sex                  11
year                  0
dtype: int64

In [5]:
df.dropna(inplace=True)

In [6]:
df['species'] = df['species'].astype('category')
df['island'] = df['island'].astype('category')
df['sex'] = df['sex'].astype('category')
df['year'] = df['year'].astype('category') 

df_normalized = df.copy()

for col in df_normalized.columns:
    if df_normalized[col].dtypes == float:
        df_normalized[col] = (df_normalized[col] - df_normalized[col].min()) / (df_normalized[col].max() - df_normalized[col].min())

In [7]:
df_normalized = pd.get_dummies(df_normalized, drop_first=True)

df_normalized.head()

bill_length_mm  bill_depth_mm  flipper_length_mm  body_mass_g  \
0        0.254545       0.666667           0.152542     0.291667   
1        0.269091       0.511905           0.237288     0.305556   
2        0.298182       0.583333           0.389831     0.152778   
4        0.167273       0.738095           0.355932     0.208333   
5        0.261818       0.892857           0.305085     0.263889   

   species_Chinstrap  species_Gentoo  island_Dream  island_Torgersen  \
0                  0               0             0                 1   
1                  0               0             0                 1   
2                  0               0             0                 1   
4                  0               0             0                 1   
5                  0               0             0                 1   

   sex_male  year_2008  year_2009  
0         1          0          0  
1         0          0          0  
2         0          0          0  
4         0          0          0  
5         1          0          0

In [8]:
Y = df_normalized[['sex_male']]
X = df_normalized.drop(['sex_male'], axis=1)

In [9]:
# Split the data into train and test
np.random.seed(8)

train_size = int(0.8 * Y.shape[0])
indis = list(range(Y.shape[0]))
np.random.shuffle(indis)
# print(indis)

train_indis = indis[:train_size]
test_indis = indis[train_size:]

y_train, y_test = Y.iloc[train_indis], Y.iloc[test_indis]
X_train, X_test = X.iloc[train_indis], X.iloc[test_indis]

X_train = X_train.to_numpy()
X_test = X_test.to_numpy()
y_train = y_train.to_numpy().reshape(-1)
y_test = y_test.to_numpy().reshape(-1)

print(X_train.shape, X_test.shape, y_train.shape, y_test.shape)

(266, 10) (67, 10) (266,) (67,)


## Building Perceptron From Scratch

In [10]:
class perceptron():
    def __init__(self, learning_rate, epochs):
        self.learning_rate = learning_rate
        self.epochs = epochs
        
    def sigmoid(self, z):
        self.z = z
        return 1 / (1 + np.exp(-self.z))
        
    def fit(self, X_train, y_train):
        self.X = np.insert(X_train, 0, 1, axis=1) # adding a column for bias term
        self.y = y_train
        self.w = np.ones(self.X.shape[1])
        # print(self.w.shape, self.X.shape)
        for i in range(self.epochs):
            self.forwardpropagate()
            self.backpropagate()
        return self.w
        
    def forwardpropagate(self):
        self.pred = self.sigmoid(np.dot(self.X, self.w))
        self.error = -(self.y - self.pred)
        # print(self.error)
        
    def backpropagate(self):
        # print(self.error * self.pred * (1 - self.pred))
        dw = np.dot(self.error * self.pred * (1 - self.pred), self.X)
        self.w = self.w - self.learning_rate * dw
    
    def predict(self, X_test):
        self.X_test = np.insert(X_test, 0, 1, axis=1)
        return np.where(self.sigmoid(np.dot(self.X_test, self.w)) >= 0.5, 1, 0)

In [11]:
model1 = perceptron(learning_rate = 1, epochs = 1000)
model1.fit(X_train, y_train)
y_pred_train = model1.predict(X_train)
y_pred_test = model1.predict(X_test)

acc_train = sum(y_pred_train == y_train)/len(y_train)
print(acc_train)

acc_test = sum(y_pred_test == y_test)/len(y_test)
print(acc_test)

0.9398496240601504
0.8805970149253731


In [12]:
model2 = perceptron(learning_rate = 0.1, epochs = 10000)
model2.fit(X_train, y_train)
y_pred_train = model2.predict(X_train)
y_pred_test = model2.predict(X_test)

acc_train = sum(y_pred_train == y_train)/len(y_train)
print(acc_train)

acc_test = sum(y_pred_test == y_test)/len(y_test)
print(acc_test)

0.9548872180451128
0.8507462686567164


In [13]:
model3 = perceptron(learning_rate = 10, epochs = 10000)
model3.fit(X_train, y_train)
y_pred_train = model3.predict(X_train)
y_pred_test = model3.predict(X_test)

acc_train = sum(y_pred_train == y_train)/len(y_train)
print(acc_train)

acc_test = sum(y_pred_test == y_test)/len(y_test)
print(acc_test)

0.48872180451127817
0.5223880597014925


In [14]:
model4 = perceptron(learning_rate = 1, epochs = 10000)
model4.fit(X_train, y_train)
y_pred_train = model4.predict(X_train)
y_pred_test = model4.predict(X_test)

acc_train = sum(y_pred_train == y_train)/len(y_train)
print(acc_train)

acc_test = sum(y_pred_test == y_test)/len(y_test)
print(acc_test)

0.9624060150375939
0.8656716417910447


## Performance and comparison with Logistic Regression model

- I used a perceptron with sigmoid activation function for the model
- For error term I took the squared error function $\frac{1}{2}(y-\hat{y})^2$ even though this is a classification problem since it's easier to calculate the derivative of this error term. Also, logistic regression can be taken as a linear model since the output depends on the linear combination of features.
- For backpropagation the calculated formula was: $\frac{\partial{E}}{\partial{w}} = -(y-\hat{y}) a (1-a) x$ where $a = \frac{1}{1+e^{-z}}$, and x corresponds to weight w

My model performance was fairly good on the Penguin dataset even though it was a single perceptron
- We can note that if the learning rate is very small, the model doesn't converge to best parameters fast enough. On the other hand a large learning rate makes the model diverge
- Similarly, when I used insufficient number of epochs the model didn't converge


For my best performing model the training accuracy was 96% and test accuracy was 86.5% (assuming that one model where test accuracy is 88% but training accuracy is just 93% is just due to a bad split)

This is only slightly worse than the logistic regression model implemented using gradient descent where the accuracy was 88%